jlm/llvm/opt/alias-analyses/AliasAnalysisPrecisionEvaluator.cpp

```diff
@@ -0,0 +1,501 @@
+    AddMeasurement(NumTotalMustAlias_, clobberInfos.TotalMustAlias);
+  }
+
+  static std::unique_ptr<PrecisionStatistics>
+  Create(const util::FilePath & sourceFile)
+  {
+    return std::make_unique<PrecisionStatistics>(sourceFile);
+  }
+};
+
+AliasAnalysisPrecisionEvaluator::AliasAnalysisPrecisionEvaluator() = default;
+
+AliasAnalysisPrecisionEvaluator::~AliasAnalysisPrecisionEvaluator() noexcept = default;
+
+void
+AliasAnalysisPrecisionEvaluator::EvaluateAliasAnalysisClient(
+    const rvsdg::RvsdgModule & rvsdgModule,
+    AliasAnalysis & aliasAnalysis,
+    util::StatisticsCollector & statisticsCollector)
+{
+  auto statistics = PrecisionStatistics::Create(rvsdgModule.SourceFilePath().value());
+
+  // If precision evaluation statistics are not demanded, skip doing it
+  if (!statisticsCollector.IsDemanded(*statistics))
+    return;
+
+  Context_ = Context{};
+
+  // If creating an aliasing graph is enabled, initialize an empty graph
+  util::graph::Writer gw;
+  if (IsAliasingGraphEnabled())
+  {
+    Context_.AliasingGraph_ = &gw.CreateGraph();
+
+    // Also emit the RVSDG to the graph writer
+    dot::LlvmDotWriter writer;
+    writer.WriteGraphs(gw, rvsdgModule.Rvsdg().GetRootRegion(), true);
+  }
+
+  // Do the pairwise alias queries
+  statistics->StartEvaluatingPrecision(*this, aliasAnalysis);
+  EvaluateAllFunctions(rvsdgModule.Rvsdg().GetRootRegion(), aliasAnalysis);
+  statistics->StopEvaluatingPrecision();
+
+  // If an aliasing graph was constructed during the evaluation, print it out now
+  if (IsAliasingGraphEnabled())
+  {
+    auto out = statisticsCollector.CreateOutputFile("AliasingGraph.dot", true);
+    std::ofstream fd(out.path().to_str());
+    gw.OutputAllGraphs(fd, util::graph::OutputFormat::Dot);
+    statistics->AddAliasingGraphOutputFile(out.path());
+  }
+
+  // If precision metrics are demanded per function, create the output file
+  std::optional<util::FilePath> perFunctionOutputFile;
+  if (IsPerFunctionOutputEnabled())
+  {
+    perFunctionOutputFile =
+        statisticsCollector.CreateOutputFile("AAPrecisionEvaluation.log", true).path();
+    statistics->AddPerFunctionOutputFile(*perFunctionOutputFile);
+  }
+
+  // Calculate total and average precision statistics
+  CalculateResults(perFunctionOutputFile, *statistics);
+
+  statisticsCollector.CollectDemandedStatistics(std::move(statistics));
+}
+
+void
+AliasAnalysisPrecisionEvaluator::EvaluateAllFunctions(
+    const rvsdg::Region & region,
+    AliasAnalysis & aliasAnalysis)
+{
+  for (auto & node : region.Nodes())
+  {
+    if (auto lambda = dynamic_cast<const rvsdg::LambdaNode *>(&node))
+    {
+      EvaluateFunction(*lambda, aliasAnalysis);
+    }
+    else if (auto structural = dynamic_cast<const rvsdg::StructuralNode *>(&node))
+    {
+      for (size_t n = 0; n < structural->nsubregions(); n++)
+      {
+        EvaluateAllFunctions(*structural->subregion(n), aliasAnalysis);
+      }
+    }
+  }
```

Comment:
This entire function is a bit wasteful, no? It traverses the entire region tree, but ultimately only cares about lambda nodes. This is a debug pass, so it ultimately does not matter. I leave it to you whether you short-circuit this function and not traverse into gamma, theta, and delta node subregions.

Reply:
You are right, I misread the code. Thanks for making it clearer though.